sxs/waveforms/alignment.py

```diff
@@ -310,3 +299,227 @@ def align2d(wa, wb, t1, t2, n_brute_force_δt=None, n_brute_force_δϕ=5, includ
+
+    Parameters
+    ----------
+    wa : WaveformModes
+    wb : WaveformModes
+        Waveforms to be aligned
+    t1 : float
+    t2 : float
+        Beginning and end of integration interval
+    n_brute_force_δt : int, optional
+        Number of evenly spaced δt values between (t1-t2) and (t2-t1) to sample
+        for the initial guess.  By default, this is just the maximum number of
+        time steps in the range (t1, t2) in the input waveforms.  If this is
+        too small, an incorrect local minimum may be found.
+    n_brute_force_δSO3 : int, optional
+        Number of evenly spaced values over the two sphere to sample
+        for the initial guess. Dy default, this is 5.
+    max_δt : float, optional
+        Max δt to allow for when choosing the initial guess.
+    max_δSO3 : float, optional
+        Max δtheta (away from z) to allow for when choosing the initial guess.
+    include_modes: list, optional
+        A list containing the (ell, m) modes to be included in the L² norm.
+    nprocs: int, optional
+        Number of cpus to use. Default is maximum number.
+        If -1 is provided, then no multiprocessing is performed.
+
+    Returns
+    -------
+    error: float
+        Cost of scipy.optimize.least_squares
+        This is 0.5 ||wa - wb||² / ||wb||²
+    wa_prime: WaveformModes
+        Resulting waveform after transforming `wa` using `optimum`
+    optimum: OptimizeResult
+        Result of scipy.optimize.least_squares
+
+    Notes
+    -----
+    Choosing the time interval is usually the most difficult choice to make when
+    aligning waveforms.  Assuming you want to align during inspiral, the times must
+    span sufficiently long that the waveforms' norm (equivalently, orbital
+    frequency changes) significantly from `t1` to `t2`.  This means that you cannot
+    always rely on a specific number of orbits, for example.  Also note that
+    neither number should be too close to the beginning or end of either waveform,
+    to provide some "wiggle room".
+
+    """
+    from scipy.interpolate import CubicSpline
+    from scipy.optimize import least_squares
+    from .. import WaveformModes
+
+    wa_orig = wa
+    wa = wa.copy()
+    wb = wb.copy()
+
+    # Check that (t1, t2) makes sense and is actually contained in both waveforms
+    if t2 <= t1:
+        raise ValueError(f"(t1,t2)=({t1}, {t2}) is out of order")
+    if wa.t[0] > t1 or wa.t[-1] < t2:
+        raise ValueError(f"(t1,t2)=({t1}, {t2}) not contained in wa.t, which spans ({wa.t[0]}, {wa.t[-1]})")
+    if wb.t[0] > t1 or wb.t[-1] < t2:
+        raise ValueError(f"(t1,t2)=({t1}, {t2}) not contained in wb.t, which spans ({wb.t[0]}, {wb.t[-1]})")
+
+    if max_δt is None:
+        max_δt = np.inf
+
+    # Figure out time offsets to try
+    δt_lower = min(max_δt, max(t1 - t2, wa.t[0] - t1))
+    δt_upper = min(max_δt, min(t2 - t1, wa.t[-1] - t2))
+
+    # We'll start by brute forcing, sampling time offsets evenly at as many
```

Comment:
I think we need to rethink the approach of this function, starting here.  The name of the game is to get good initial guesses, but the curse of dimensionality means that brute-force methods are that much worse.   So it would probably be better to optimize hierarchically.  I think the way to go would be to

  1. Run `align1d` to get the initial guess for the time offset `δt`.  The cost function here is invariant under rotations, so the crazy precession that this function should handle shouldn't be an problem.  And I would guess that `δt` will be a great guess for the final result from this function.
  2. Apply that `δt`, then measure the *unit* angular velocity vector for each — within (t1,t2) — and use `quaternionic.align` (only available in v1.0.11) to get the rotor `R` that will best align their angular velocities.  For precessing systems, that could be enough to fully align them, but either way, there's no harm in checking the rotation about that axis.
  3. Using `δt` and `R`, brute-force optimize `R*exp(theta*z/2)` for `4*ell_max+2` values of `theta` in [0,2*pi).
  4. Use those values as the starting point for the active 4D optimization, which should be pretty easy and quick.